python/ray/tune/tests/test_multinode_sync.py

```diff
@@ -161,6 +165,143 @@ def on_step_begin(self, iteration, trials, **info):
             local_dir="/tmp/ray_results/",
         )
 
+    def testCheckpointSync(self):
+        """Test that checkpoints are correctly synced."""
+        self.cluster.update_config(
+            {
+                "provider": {
+                    "head_resources": {"CPU": 4, "GPU": 0},
+                    "env_vars": {"TUNE_GLOBAL_CHECKPOINT_S": "0"},
+                },
+                "available_node_types": {
+                    "ray.worker.cpu": {
+                        "resources": {"CPU": 4},
+                        "min_workers": 2,
+                        "max_workers": 2,
+                    },
+                    "ray.worker.gpu": {
+                        "min_workers": 0,
+                        "max_workers": 0,  # No GPU nodes
+                    },
+                },
+            }
+        )
+        self.cluster.start()
+        self.cluster.connect(client=True, timeout=120)
+        self.cluster.wait_for_resources({"CPU": 12})
+
+        def train(config, checkpoint_dir=None):
+            start = 0
+            if checkpoint_dir:
+                with open(os.path.join(checkpoint_dir, "checkpoint.json"), "rt") as fp:
+                    state = json.load(fp)
+                    start = state["step"] + 1
+
+            for i in range(start, start + 10):
+                with tune.checkpoint_dir(i) as d:
+                    with open(os.path.join(d, "checkpoint.json"), "wt") as fp:
+                        json.dump({"step": i}, fp)
+                tune.report(step=i)
+
+            with open(f"/cluster/shared/indicator.{tune.get_trial_id()}", "wt") as fp:
+                fp.write("")
+
+            time.sleep(6)
+            tune.report(step=i + 1)
```

Comment:
i is out of scope here.
But is this just to let the function run a bit while the driver is gathering enough "indicator" files to crash?

Reply:
That's right. `i` is guaranteed to be set within the test setting, but I've specifically initialized it now to make the IDE complaint go away.